Adopt the ROKS workaround for Tigera Operator

SzabolcsSzoke opened the pull request:
The new ROKS versions are using Tigera Operator which reads the Installation resource and it changes the IPPools' IPIPMode settings according to the Installation.
So move this ROKS workaround from default IPPool into default Installation

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our developer guide: https://submariner.io/development/
2. Ensure you have added the appropriate tests for your PR: https://submariner.io/development/code-review/#test-new-functionality
3. Read the code review guide to ease the review process: https://submariner.io/development/code-review/
4. If the PR is unfinished, mark it as a draft: https://submariner.io/development/code-review/#mark-work-in-progress-prs-as-drafts
5. If you are using CI to debug, use your private fork: https://submariner.io/development/code-review/#use-private-forks-for-debugging-prs-by-running-ci
6. Add labels to the PR as appropriate.

This template is based on the K8s/K8s template:

https://github.com/kubernetes/kubernetes/blob/master/.github/PULL_REQUEST_TEMPLATE.md
-->

SzabolcsSzoke committed:
Adopt the ROKS workaround for Tigera Operator

The new ROKS versions are using Tigera Operator which reads the
Installation resource and it changes the IPPools' IPIPMode settings
according to the Installation.
So move this ROKS workaround from default IPPool into default Installation.

tpantelis commented on pkg/routeagent_driver/handlers/calico/calico_suite_test.go:
```diff
@@ -32,27 +32,53 @@ import (
+				},
+			},
+		})
+		t.dynClient = dynamicfake.NewSimpleDynamicClient(runtime.NewScheme(), installation)
```
We can use the actual `Installation` type here instead of `Unstructured`:

```
import "k8s.io/client-go/kubernetes/scheme"
...
		tigerav1.AddToScheme(scheme.Scheme)

		t.dynClient = dynamicfake.NewSimpleDynamicClient(scheme.Scheme, &tigerav1.Installation{
			TypeMeta: metav1.TypeMeta{},
			ObjectMeta: metav1.ObjectMeta{
				Name: "default",
			},
			Spec: tigerav1.InstallationSpec{
				CalicoNetwork: &tigerav1.CalicoNetworkSpec{
					IPPools: []tigerav1.IPPool{
						{
							Encapsulation: tigerav1.EncapsulationIPIPCrossSubnet,
						},
					},
				},
			},
		})
```

SzabolcsSzoke replied:
Changed. Thx

tpantelis commented on pkg/routeagent_driver/handlers/calico/calico_suite_test.go:
```diff
@@ -135,13 +161,40 @@ func (t *testDriver) getDefaultIPPoolIPIPMode() string {
+func (t *testDriver) getDefaultInstallationEncapsulation() string {
+
+	installation, err := calico.GetDefaultInstallation(t.dynClient, gvrInstallations)
+	Expect(err).To(Succeed())
```
For sanity check to avoid potential panic:

```suggestion
	Expect(err).To(Succeed())
	Expect(installation.Spec.CalicoNetwork).NotTo(BeNil())
	Expect(installation.Spec.CalicoNetwork.IPPools).NotTo(BeEmpty())
```

SzabolcsSzoke replied:
Added the extra checks.

tpantelis commented on pkg/routeagent_driver/handlers/calico/ippool_handler.go:
```diff
@@ -66,11 +70,23 @@ var NewClient = func(restConfig *rest.Config) (calicocs.Interface, error) {
 var logger = log.Logger{Logger: logf.Log.WithName("CalicoIPPool")}
 
-func NewCalicoIPPoolHandler(restConfig *rest.Config, namespace string, k8sClient clientset.Interface) event.Handler {
+var gvrInstallations = schema.GroupVersionResource{
```
We can make this public and share with unit tests:

```suggestion
var InstallationsGVR = schema.GroupVersionResource{
```

tpantelis replied:
Also similarly for the default `Installation` name:

`DefaultInstallationName = "default"`

SzabolcsSzoke replied:
`gvrInstallations` renamed/changed to public.
Added `DefaultInstallationName`
Using them everywhere.

tpantelis commented on pkg/routeagent_driver/handlers/calico/calico_suite_test.go:
```diff
@@ -135,13 +161,40 @@ func (t *testDriver) getDefaultIPPoolIPIPMode() string {
-	}, metav1.CreateOptions{})
+	}
+
+	_, err := t.dynClient.Resource(gvrService).Namespace(eventtesting.Namespace).Create(context.Background(), service, metav1.CreateOptions{})
```
Note that `annotations` is unused in the code above:

As before, we can use the actual type here:

```
test.CreateResource(t.dynClient.Resource(corev1.SchemeGroupVersion.WithResource("services")).Namespace(eventtesting.Namespace),
		&corev1.Service{
			ObjectMeta: metav1.ObjectMeta{
				Name:        calico.GwLBSvcName,
				Annotations: annotations,
			},
		})
```

SzabolcsSzoke replied:
Rewritten it to use actual type as suggested. Thx

tpantelis commented on pkg/routeagent_driver/handlers/calico/calico_suite_test.go:
```diff
@@ -135,13 +161,40 @@ func (t *testDriver) getDefaultIPPoolIPIPMode() string {
+func (t *testDriver) patchCalicoNetworkIpPoolsEncapsulation() {
+	err := calico.PatchCalicoNetworkIpPoolsEncapsulation(t.dynClient, gvrInstallations, tigerav1.EncapsulationIPIPCrossSubnet.String(), tigerav1.EncapsulationIPIP.String())
+	Expect(err).To(Succeed())
+}
```
Ideally we should avoid sharing production code with the unit test code as it's the production code we're trying to verify. In this case, instead of patching the resource in the tests, we can create the resource with the appropriate fields on test setup, which it's already doing. AFAICT, this patch update isn't even needed since the encapsulation field is already set to `EncapsulationIPIPCrossSubnet` on creation in the test setup. So this function can be removed.

SzabolcsSzoke replied:
This function was deleted, because doesn't needed anymore.

tpantelis commented on pkg/routeagent_driver/handlers/calico/calico_suite_test.go:
```diff
@@ -135,13 +161,40 @@ func (t *testDriver) getDefaultIPPoolIPIPMode() string {
 
+func (t *testDriver) getDefaultInstallationEncapsulation() string {
+
+	installation, err := calico.GetDefaultInstallation(t.dynClient, gvrInstallations)
```
As with `patchCalicoNetworkIpPoolsEncapsulation` below, let's avoid using the production code here:

```suggestion
	installation := test.GetResource(t.dynClient.Resource(gvrInstallations), &tigerav1.Installation{ObjectMeta: metav1.ObjectMeta{
		Name: "default",
	}})
```

SzabolcsSzoke replied:
Changed.Thx

tpantelis commented on pkg/routeagent_driver/handlers/calico/ippool_handler.go:
```diff
@@ -240,65 +299,47 @@ func (h *calicoIPPoolHandler) updateROKSCalicoCfg(ctx context.Context) error {
+	err = PatchCalicoNetworkIpPoolsEncapsulation(h.dynClient, gvrInstallations, encapsulation, tigerav1.EncapsulationIPIP.String())
+	if err != nil {
+		return err
+	}
```
I think we should use `util.Update` here as the prior code did. It does the retrieval and only updates if something changed. It also handles conflicts and retries. It doesn't use patch but I don't think that's significant here (plus patch is typically used when updating one field anyway).  Then you can remove `GetDefaultInstallation` and `PatchCalicoNetworkIpPoolsEncapsulation`.

Also use `util.Update` similarly in `restoreROKSCalicoCfg`.

```
    return util.Update(ctx, resource.ForDynamic(h.dynClient.Resource(tigerav1.GroupVersion.WithResource("installations"))),
        resource.MustToUnstructured(&tigerav1.Installation{
		    ObjectMeta: metav1.ObjectMeta{
		         Name: "default",
	         },
	}), func(existing *unstructured.Unstructured) (*unstructured.Unstructured, error) {
		installation := &tigerav1.Installation{}
		err := runtime.DefaultUnstructuredConverter.FromUnstructured(existing.Object, installation)
		utilruntime.Must(err)

		if installation.Spec.CalicoNetwork == nil {
			installation.Spec.CalicoNetwork = &tigerav1.CalicoNetworkSpec{}
		}

		ipPools := installation.Spec.CalicoNetwork.IPPools
		if len(ipPools) > 0 && ipPools[0].Encapsulation == tigerav1.EncapsulationIPIP {
			logger.Infof("Encapsulation of default Installations already set to IPIP")
			return existing, nil
		}

		if installation.Annotations == nil {
			installation.Annotations = map[string]string{}
		}

		installation.Annotations[submarinerPrevEncapsulation] = ipPools[0].Encapsulation.String()

		if len(ipPools) > 0 {
			ipPools[0].Encapsulation = tigerav1.EncapsulationIPIP
		} // Should add an IPPool if none exists?

		return resource.MustToUnstructured(installation), nil
	})
```

SzabolcsSzoke replied:
Rewritten as suggested. Thx

tpantelis commented on pkg/routeagent_driver/handlers/calico/ippool_handler.go:
```diff
@@ -240,65 +299,47 @@ func (h *calicoIPPoolHandler) updateROKSCalicoCfg(ctx context.Context) error {
-		UpdateFunc: h.client.ProjectcalicoV3().IPPools().Update,
-		DeleteFunc: h.client.ProjectcalicoV3().IPPools().Delete,
+	encapsulationPatch := []byte(`{"op": "replace", "path": "/spec/calicoNetwork/ipPools/0/encapsulation", "value": "` + prevEncapsulation + `"}`)
+	labelPatch := []byte(`{"op": "remove", "path": "/metadata/labels/` + submarinerManagedLabel + `"}`)
```
This is set and removed on restore but otherwise unused AFAICT. The `submarinerPrevEncapsulation` annotation indicates if it was  modified so I don't think we need the `submarinerManagedLabel`.

SzabolcsSzoke replied:
The `submarinerManagedLabel` is coming from the previous workaround and it is used internally (IBM's private code) to identify and handle clusters with Submariner.

tpantelis commented on pkg/routeagent_driver/handlers/calico/ippool_handler.go:
```diff
@@ -225,6 +273,17 @@ func (h *calicoIPPoolHandler) platformIsROKS(ctx context.Context) (bool, error)
+	if err != nil {
+		return false, err
+	}
+
```
You don't need to convert to `Service` here as `Unstructured` implements the `ObjectMeta` interface, ie `serviceUnstructured.GetAnnotations()` ...

SzabolcsSzoke replied:
Thx. Removed unnecessary conversion.

tpantelis commented on pkg/routeagent_driver/handlers/calico/ippool_handler.go:
```diff
@@ -214,9 +230,41 @@ func getEndpointSubnetIPPoolName(endpoint *submV1.Endpoint, subnet string) strin
 func (h *calicoIPPoolHandler) platformIsROKS(ctx context.Context) (bool, error) {
 	// Submariner GW is deployed on ROKS using LB service with specific annotations.
-	service, err := h.k8sClient.CoreV1().Services(h.namespace).Get(ctx, GwLBSvcName, metav1.GetOptions{})
+	serviceUnstructured, err := h.dynClient.Resource(gvrService).Namespace(h.namespace).Get(ctx, GwLBSvcName, metav1.GetOptions{})
```
```suggestion
	serviceUnstructured, err := h.dynClient.Resource(corev1.SchemeGroupVersion.WithResource("services")).Namespace(h.namespace).Get(ctx, GwLBSvcName, metav1.GetOptions{})
```

SzabolcsSzoke replied:
Changed. Thx

tpantelis commented on pkg/routeagent_driver/handlers/calico/ippool_handler.go:
```diff
@@ -66,11 +76,17 @@ var NewClient = func(restConfig *rest.Config) (calicocs.Interface, error) {
+	Group:    "operator.tigera.io",
+	Version:  "v1",
+	Resource: "installations",
+}
```
```suggestion
var InstallationsGVR = tigerav1.GroupVersion.WithResource("installations")
```

SzabolcsSzoke replied:
Changed.

tpantelis commented on pkg/routeagent_driver/handlers/calico/ippool_handler.go:
```diff
@@ -240,65 +277,62 @@ func (h *calicoIPPoolHandler) updateROKSCalicoCfg(ctx context.Context) error {
-	return errors.Wrap(err, "error updating default ippool")
+	// platform is ROKS, make sure that encapsulation of default Installation is set to IPIP
+	// In this way the Tigera Operator will set IPIPMode of default IPPool to Always
+	return util.Update(ctx, resource.ForDynamic(h.dynClient.Resource(tigerav1.GroupVersion.WithResource("installations"))),
```
```suggestion
	return util.Update(ctx, resource.ForDynamic(h.dynClient.Resource(InstallationsGVR)),
```

SzabolcsSzoke replied:
Changed.

tpantelis commented on pkg/routeagent_driver/handlers/calico/ippool_handler.go:
```diff
@@ -240,65 +277,62 @@ func (h *calicoIPPoolHandler) updateROKSCalicoCfg(ctx context.Context) error {
+	if err != nil {
+		return err
 	}
+
```
Please use `util.Update` here as well. Then you can also remove `GetDefaultInstallation`.

SzabolcsSzoke replied:
Rewritten to use `util.Update` and removed `GetDefaultInstallation`. Thx

tpantelis commented:
Also:

- Please squash the 2 commits (we don't include commits addressing PR comments in the commit history)
- Need to address conflicts as indicated below
- Need to sign your commits
- The summary title in https://github.com/submariner-io/submariner/pull/3570/commits/3e3d0cbef34255ea0030c46ae033525151438fda is too long (the limit's around 70 chars).

SzabolcsSzoke commented:
> Also:
> 
> * Please squash the 2 commits (we don't include commits addressing PR comments in the commit history)
> * Need to address conflicts as indicated below
> * Need to sign your commits
> * The summary title in [3e3d0cb](https://github.com/submariner-io/submariner/commit/3e3d0cbef34255ea0030c46ae033525151438fda) is too long (the limit's around 70 chars).

Hi,

* I squash my commits into one.
* Handle the conflicts
* Used shorter summary title
* And signed my commit
Thx.

tpantelis commented on pkg/routeagent_driver/handlers/calico/ippool_handler.go:
```diff
@@ -240,65 +249,66 @@ func (h *calicoIPPoolHandler) updateROKSCalicoCfg(ctx context.Context) error {
+
+			if len(ipPools) > 0 {
+				ipPools[0].Encapsulation = tigerav1.EncapsulationIPIP
+			} // Should add an IPPool if none exists?
```
Sorry, just one more thing here... I had added this comment in my suggestion to raise awareness. If no pool exists and we don't need to add one, then we should also skip adding the annotations as well, so above:

```
            ipPools := installation.Spec.CalicoNetwork.IPPools
			if len(ipPools) == 0 {
                 logger.Infof("No IPPool found for Installation %q", DefaultInstallationName)
				return existing, nil
             } else if ipPools[0].Encapsulation == tigerav1.EncapsulationIPIP {
				logger.Infof("Encapsulation of %q Installation already set to IPIP", DefaultInstallationName)
				return existing, nil
			}
```

```suggestion
		    ipPools[0].Encapsulation = tigerav1.EncapsulationIPIP
```

SzabolcsSzoke replied:
Changed.

tpantelis commented:
[Your commit](https://github.com/submariner-io/submariner/pull/3570/commits/2abc597de92110844b6652567bce730a08925d8a) still isn't signed.

tpantelis commented on pkg/routeagent_driver/handlers/calico/ippool_handler.go:
```diff
@@ -240,65 +253,64 @@ func (h *calicoIPPoolHandler) updateROKSCalicoCfg(ctx context.Context) error {
 
-		return existing, nil
-	})
+			ipPools[0].Encapsulation = tigerav1.EncapsulationIPIP
```
This will panic if `ipPools` is empty.

SzabolcsSzoke replied:
Thx. Added a precheck and if `ipPools` is empty will just return.

SzabolcsSzoke committed:
Merge branch 'devel' into changeCalicoInstallation4ROKS

Signed-off-by: SzabolcsSzoke <88749064+SzabolcsSzoke@users.noreply.github.com>

tpantelis reviewed (approved):
The DCO job is still failing.

mihivagyok commented:
/ready-to-test

skitt committed:
Merge branch 'devel' into changeCalicoInstallation4ROKS
